test-data/unit/fine-grained.test

```diff
@@ -629,6 +629,252 @@ class M(type):
 ==
 a.py:4: error: Incompatible types in assignment (expression has type "str", variable has type "int")
 
+[case testDataclassUpdate1]
+# flags: --python-version 3.7
+[file a.py]
+from dataclasses import dataclass
+
+@dataclass
+class A:
+    x: int
+
+[file b.py]
+from dataclasses import dataclass
+
+from a import A
+@dataclass
+class B(A):
+    y: int
+
+B(1, 2)
+
+[file a.py.2]
+from dataclasses import dataclass
+
+@dataclass
+class A:
+    x: str
+
+[out]
+==
+b.py:8: error: Argument 1 to "B" has incompatible type "int"; expected "str"
+[builtins fixtures/list.pyi]
+
+[case testDataclassUpdate2]
+# flags: --python-version 3.7
+[file c.py]
+Foo = int
+
+[file c.py.2]
+Foo = str
+
+[file a.py]
+from dataclasses import dataclass
+from c import Foo
+
+@dataclass
+class A:
+    x: Foo
+
+[file b.py]
+from dataclasses import dataclass
+
+from a import A
+@dataclass
+class B(A):
+    y: int
+
+B(1, 2)
+
+[out]
+==
+b.py:8: error: Argument 1 to "B" has incompatible type "int"; expected "str"
+[builtins fixtures/list.pyi]
+
+[case testDataclassUpdate3]
+# flags: --python-version 3.7
+from b import B
+B(1, 2)
+[file b.py]
+from a import A
+from dataclasses import dataclass
+@dataclass
+class B(A):
+    b: int
+[file a.py]
+from dataclasses import dataclass
+@dataclass
+class A:
+    a: int
+
+[file a.py.2]
+from dataclasses import dataclass
+@dataclass
+class A:
+    a: int
+    other: int
+[builtins fixtures/list.pyi]
+[out]
+==
+main:3: error: Too few arguments for "B"
+
+[case testDataclassUpdate4]
+# flags: --python-version 3.7
+from b import B
+B(1, 2)
+[file b.py]
+from a import A
+from dataclasses import dataclass
+@dataclass(frozen=True)
+class B(A):
+    b: int
+[file a.py]
+from dataclasses import dataclass
+@dataclass(frozen=True)
+class A:
+    a: int
+
+[file a.py.2]
+from dataclasses import dataclass
+@dataclass(frozen=True)
+class A:
+    a: int
+    other: int
+[builtins fixtures/list.pyi]
+[out]
+==
+main:3: error: Too few arguments for "B"
+
+[case testDataclassUpdate5]
+# flags: --python-version 3.7
+from b import B
+B(1, 2)
+[file b.py]
+from a import A
+from dataclasses import dataclass
+@dataclass
+class B(A):
+    b: int
+[file a.py]
+from dataclasses import dataclass
+@dataclass(init=False)
+class A:
+    a: int
+
+[file a.py.2]
+from dataclasses import dataclass
+@dataclass(init=False)
+class A:
+    a: int
+    other: int
+[builtins fixtures/list.pyi]
+[out]
+==
+main:3: error: Too few arguments for "B"
```

Comment:
Another idea for test is to have more than one update. For example, introduce this error, and then check that it disappears again when you delete another attribute.